### PR TITLE
feat: Add ability to delete branches (#361)

### LINKED ISSUE #361: [FEATURE]: Delete branches
### Description

Branches cannot be deleted ATM.

### Possible solution

I love how github does it, lets implement similar.

image

### Alternatives

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ttebify**
# Description

> Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #361

## Type of change

> Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I documented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: apps/platform/trpc/routes/branch.ts [hunk 1]
```diff
@@ -158,4 +162,53 @@ export const branches = createRouter({
 
       return updatedBranch;
     }),
+  delete: withAuth
+    .input(z.object({ branchId: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      const { branchId } = input;
+
+      try {
```
Comment (chetannn) [inline apps/platform/trpc/routes/branch.ts:170]: First check whether the given branch is `protected` or not.

File: apps/platform/trpc/routes/branch.ts [hunk 2]
```diff
@@ -158,4 +162,53 @@ export const branches = createRouter({
...
+      }
+    }),
+  restoreBranch: withAuth
+    .input(
+      z.object({
+        branchId: z.string(),
+      }),
+    )
+    .mutation(async ({ input, ctx }) => {
+      const { branchId } = input;
```
Comment (chetannn) [inline apps/platform/trpc/routes/branch.ts:195]: Before `update` we should check If the branch is actually deleted or not. Just to be on the safe side.

File: apps/platform/pages/projects/[slug]/branches.tsx [hunk 3]
```diff
@@ -56,6 +59,10 @@ export const BranchesPage = ({
 }: Props) => {
   const [isOpen, setIsOpen] = useState(false);
   const [isPrModalOpen, setIsPrModalOpen] = useState(false);
+  // Keep track of deletedAt status for each row
+  const [deletedRows, setDeletedRows] = useState<{ [id: string]: boolean }>({});
```
Comment (chetannn) [inline apps/platform/pages/projects/[slug]/branches.tsx:63]: Do we really need to track `deleted branches` ?
Just refreshing the props will do the job I think.

File: apps/platform/lib/jobs/deleteBranchJob.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { Job, PrismaTransaction } from "./baseJob";
+
+export class DeleteBranchJob extends Job {
+  constructor() {
+    super();
+  }
+
+  async handle(tx: PrismaTransaction) {
+    // Get all branches that have deletedAt within the past 24 hours.
+    const branches = await tx.branch.findMany({
```
Comment (chetannn) [inline apps/platform/lib/jobs/deleteBranchJob.ts:10]: Here, we need to remove all other associated data as well like `secrets` and `pull requests`.

File: apps/platform/pages/projects/[slug]/branches.tsx (hunk 3 above)
Comment (ttebify) [reply apps/platform/pages/projects/[slug]/branches.tsx:63]: Refreshing the page will resolve the issue. This particular state is used to temporarily hold the deleted status after the user clicks on delete, and to render a restore status afterward.

**Review by dahal — CHANGES_REQUESTED** [review]
@ttebify lets rebase and use [store/Branches.ts](https://github.com/envless/envless/blob/main/apps/platform/store/Branches.ts)

Ref #363

**Pull request closed without merging**